### ISSUE TITLE
expressionPrecedence broken for Operation_Binary in UBPF p4c backend

### ISSUE THREAD

**Issue by hesingh** [p4lang/p4c#2670]
I have attached a P4 program in this issue.

The following line in the P4 code causes an interesting C code generated by the UBPF backend. The issue should be reproducible even with the EBPF backend.

```p4
headers.ipv6.ip_version_traffic_class_and_flow_label[31:12] = 52;
```
The generated C code in included below.

```c
 headers.ipv6.ip_version_traffic_class_and_flow_label = headers.ipv6.ip_version_traffic_class_and_flow_label & ~4294963200 | (uint32_t)52 << 12 & 4294963200;
```
Issue are:

First, gcc fails compiling unless parenthesis are added as follows:

 headers.ipv6.ip_version_traffic_class_and_flow_label = (headers.ipv6.ip_version_traffic_class_and_flow_label & ~4294963200) | ((uint32_t)52 << 12 & 4294963200);

`p4c/backends/ebpf/codeGen.cpp` Operation_Binary preorder needs change and maybe other methods.

Second, there are junk numbers after the `&` operator.

This is how I test:

```bash
p4c/build/$ ./p4c-ubpf --std p4-16 ../testdata/p4_16_samples/op_bin.p4 -o tst.c
```

[op_bin-p4.txt](https://github.com/p4lang/p4c/files/5968340/op_bin-p4.txt)

**Comment by hesingh** [comment]
On this line of p4c code:

https://github.com/p4lang/p4c/blob/master/backends/ebpf/codeGen.cpp#L53

`prec` is 1 and `b->getPrecedence()` is 8 and no parenthesis are added but needed. 

Of course, these code changes to p4c fix the parenthesis issue:

```c
diff --git a/backends/ebpf/codeGen.cpp b/backends/ebpf/codeGen.cpp
index 3e84cb1ba..d89319f7d 100644
--- a/backends/ebpf/codeGen.cpp
+++ b/backends/ebpf/codeGen.cpp
@@ -51,7 +51,7 @@ bool CodeGenInspector::preorder(const IR::Operation_Binary* b) {
     widthCheck(b);
     int prec = expressionPrecedence;
     bool useParens = prec > b->getPrecedence();
-    if (useParens)
+    //    if (useParens)
         builder->append("(");
     visit(b->left);
     builder->spc();
@@ -59,7 +59,7 @@ bool CodeGenInspector::preorder(const IR::Operation_Binary* b) {
     builder->spc();
     expressionPrecedence = b->getPrecedence() + 1;
     visit(b->right);
-    if (useParens)
+    //    if (useParens)
         builder->append(")");
     expressionPrecedence = prec;
     return false;
```

With the changes above, I ran `make check` and no test fails.

**Comment by mbudiu-vmw** [comment]
Feel free to submit a PR. I don't see a good reason we would not parenthesize expressions in the generated C.

**Comment by hesingh** [comment]
The gcc compiler fails with an error.   This is why I removed the `useParens` check.  I cannot issue a PR yet because the root cause to the other issue is not found - why is there junk after `&`?  Note, in my private version of VPP C code generator I wrote, I don't see the junk  

This is the failure with `useParens` is used.

```c
upf.c:1456:103: error: '&' within '|' [-Werror,-Wbitwise-op-parentheses]
hdr->ipv6.ip_version_traffic_class_and_flow_label = hdr->ipv6.ip_version_traffic_class_and_flow_label & ~0xfffff000 | (uint32_t)(uint32_t)meta->_acl_qfi29 << 12 & 0xfffff000;                                    }
                                                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~ ~
upf.c:1456:103: note: place parentheses around the '&' expression to silence this warning
hdr->ipv6.ip_version_traffic_class_and_flow_label = hdr->ipv6.ip_version_traffic_class_and_flow_label & ~0xfffff000 | (uint32_t)(uint32_t)meta->_acl_qfi29 << 12 & 0xfffff000;                                    }
                                                                                                      ^
                                                    (                                                              )
```

**Comment by hesingh** [comment]
The junk is correct if converted to hex.  It's `0xfffff000`.  So it's a minor issue.  I converted the UBPF program to a EBPF program because EBPF lets me create `const entries` inside the P4 program but UBPF does not.  I could reproduce the issue with using const entries. I have narrowed down the issue of decimal number instead of hex to the line of code below.

https://github.com/p4lang/p4c/blob/master/backends/ebpf/codeGen.cpp#L30

I will issue a PR when I get a chance - thanks.
